### ISSUE TITLE
Is implementing exponential backoff a valid use case for message.drop()?

### ISSUE THREAD

**Issue by chrisfrank** [googleapis/python-pubsub#60]
First, thanks for such a stable and robust client library. It's powering much of The New York Times photo archive :)

Our team is trying to implement exponential backoff when retrying `StreamingPullFuture` callbacks that have failed unexpectedly. One solution under consideration relies on the `drop()` method of `google.cloud.pubsub_v1.subscriber.message.Message` — but the docs [warn against using this method](https://googleapis.dev/python/pubsub/latest/subscriber/api/message.html?highlight=drop#google.cloud.pubsub_v1.subscriber.message.Message.drop),  so @michaelplaing and I were wondering if the following (abbreviated) implementation is a valid use case:

```python
 def retry(msg: pubsub_v1.subscriber.message.Message) -> None:
        delay_seconds = calculate_delay_seconds_from_failure_count(msg)  # omitted
        msg.modify_ack_deadline(delay_seconds)
        msg.drop()
```

Without the call to `msg.drop()`, the client seems to block this thread until the new ack deadline arrives. We would, ideally, like to release the thread so that the client's Scheduler can continue to dispatch other messages to it. It seems like `drop()` is doing the trick — we just want to make sure this is the expected behavior and safe to rely on. Thanks!

**Comment by plamut** [comment]
@chrisfrank Thanks for the compliments, good to hear that the library serves your needs well!

> Without the call to `msg.drop()`, the client seems to block this thread until the new ack deadline arrives. We would, ideally, like to release the thread so that the client's Scheduler can continue to dispatch other messages to it.

Which thread are you referring to, the thread used to invoke the message received callback I presume? There are two things that can affect this:
- **The [flow control](https://cloud.google.com/pubsub/docs/pull#config) settings** limit the number of outstanding messages. If a subscriber cannot process the messages fast enough, the streaming pull temporarily suspends pulling new messages, and only resumes the stream when some of the messages are processed and the load decreases.

  The load is computed [here](https://github.com/googleapis/python-pubsub/blob/a3ee9413409fade06ec218a4268633592a3d4dec/google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py#L209-L242), and is basically proportional to the number messages in lease management. Dropping a message from the lease management decreases the load, which can result in resuming the stream and new messages delivered.

  However, the backend itself will not know anything about it and will not resend the message until the ACK deadline is hit. If an error occurs in the callback, it is thus better to NACK the message (`msg.nack()`) - in addition to dropping the message from the lease management, it will also immediately expire the message on the backend.

 - The **callback scheduler thread pool size** - this is used to asynchronously dispatch received messages to the user-provided callback. By default, the maximum number of workers is 10.

   If the user callback takes a long time to process each message and blocks the thread, and all scheduler threads are are busy, the scheduler will not dispatch additional messages even if the current load is still below the flow control limits.

   A way around this is to either provide a custom scheduler to `subscriber.subscribe()` with a higher max worker count, or to move the heavy-lifting out of the message callback - the callback can simply place the message into a common queue and return, freeing the scheduler thread, and the messages in the queue are actually processed by a different user thread/thread pool.

Does any of this help with your problem with blocked threads?

If not or if I misunderstood the question, please elaborate on what kind of thread blocking are you experiencing, thanks!

**Comment by pradn** [comment]
Hey Chris! Glad you are getting some good use out of this library at the NYT.

The StreamingPullFuture fails for two reasons: 1) when the underlying streaming pull connection breaks for some unrecoverable reason 2) the user callback throws an exception. Are you trying to delay the re-delivery of messages that were delivered and not acked at the time of the StreamingPullFuture failure?

Peter's answer above is excellent. I'd like to add a few points to it.

The reason why drop() is unblocking the thread is because it reduces the number of messages under lease management by the subscriber. This reduces the load factor (more precisely, the numerator [here](https://github.com/googleapis/python-pubsub/blob/master/google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py#L237)), causing the subscriber to release some messages.

It's okay to use drop() for this use-case. In fact, you must. If you don't drop() the message, the subscriber will keep leasing the message until the maximum lease duration. So, your modify_ack_deadline() will be overwritten by subsequent mod-acks by the subscriber. (The subscriber keeps mod-acking the message until either it hits the maximum lease duration or the message is dropped.) If you want the message to be resent by CPS after your custom retry duration, you must prevent the leaser from messing with your ack-deadlines.

(By the way, there's an upcoming [dead-letter queues feature](https://github.com/googleapis/python-pubsub/blob/master/google/cloud/pubsub_v1/proto/pubsub.proto#L682) you might want to look into. It's available for whitelisted customers in alpha.)

**Comment by pradn** [comment]
We area also working on a RetryPolicy feature. It'll let you set the behavior of exponential backoff for message redeliveries. (You can set the min and max delays, with exponential backoff in between.) We can't say when it'll be available just yet.

**Comment by chrisfrank** [comment]
Thanks so much, @plamut and @pradn — this is very helpful information!

@plamut, you're correct that I was referring to the thread used to invoke the "message received" callback. We're trying to set a custom retry delay after our callback raises an exception — perhaps because a DB or API is down, or because of a bug in our code, etc. After reading your reply, I suspect our callback scheduler's ThreadPoolExecutor did have threads available, and that the "blocking" was caused by the stream being temporarily suspended. We keep FlowControl's `max_messages` value pretty low to prevent our subscribers from hoarding more messages than they can process. (Processing these messages requires doing work on large image files, and it can take a while.)

When you say...

> If an error occurs in the callback, it is thus better to NACK the message (msg.nack())

Do you mean we should call `nack()` instead of `modify_ack_deadline(delay)` followed by `drop()`? In testing, the behavior we were seeing from `nack()` is that the CPS would immediately redeliver the message — whereas with "modify" + "drop", CPS seems to respect the delay we set before it redelivers.

@pradn it sounds like we'll eventually be able to replace this implementation with the forthcoming `RetryPolicy`, which is great news! Until then, my read of your reply is that using modify + drop should be okay in the interim. Do I have that right?

Please let me know if I've misunderstood anything, and again, thank you both for such helpful responses.

**Comment by plamut** [comment]
> Do you mean we should call `nack()` instead of `modify_ack_deadline(delay)` followed by `drop()`?

Actually, the sample is fine as it is, I mentioned `nack()` because I misunderstood the "retrying the callback" part. I now see that you actually want to exponentially postpone the message re-delivery itself, while I assumed you wanted to get the message again ASAP, but postpone its processing.

As @pradn said, the upcoming RetryPolicy feature will allow that, but in the meantime the workaround with extend deadline + drop message is probably fine. Just keep in mind that might be a race condition lurking:
```py
msg.modify_ack_deadline(delay_seconds)
# the leaser thread might modify the ACK deadline set above just here
msg.drop()
```
> I suspect our callback scheduler's ThreadPoolExecutor did have threads available, and that the "blocking" was caused by the stream being temporarily suspended. 

If you want to check, you can temporarily set the logging level to DEBUG and check the records for information on [pausing](https://github.com/googleapis/python-pubsub/blob/a3ee9413409fade06ec218a4268633592a3d4dec/google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py#L274-L279) / [resuming](https://github.com/googleapis/python-pubsub/blob/a3ee9413409fade06ec218a4268633592a3d4dec/google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py#L303-L307) the message stream, as well as the current load.

**Comment by chrisfrank** [comment]
Understood! I'll go ahead and close this issue. Thank you both very much for your help!

**Comment by pradn** [comment]
@chrisfrank The [API](https://github.com/googleapis/googleapis/blob/master/google/pubsub/v1/pubsub.proto#L707) for RetryPolicy has been made public, so you can see how it looks like.

**Comment by maheshgattani** [comment]
You can now configure a retry policy in your Cloud Pub/Sub subscription. Retry policy allows users to have per message exponential backoff in case of failures.

More information can be found [here](https://cloud.google.com/pubsub/docs/admin).

**Comment by chrisfrank** [comment]
Thanks for the update, @maheshgattani! We're going to give this a try this week.
